plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3ExchangeStorage.java

```diff
@@ -0,0 +1,686 @@
+        return new S3ExchangeSliceOutput(s3, bucketName, key, uploadIdFactory, uploadExecutor, streamingUploadPartSize, sseCustomerKey);
+    }
+
+    private InitiateMultipartUploadResult initMultipartUpload(String bucket, String key, Optional<SSECustomerKey> sseCustomerKey)
+    {
+        InitiateMultipartUploadRequest request = new InitiateMultipartUploadRequest(bucket, key)
+                .withObjectMetadata(new ObjectMetadata())
+                .withCannedACL(CannedAccessControlList.Private);
+        sseCustomerKey.ifPresent(request::withSSECustomerKey);
+
+        return s3.initiateMultipartUpload(request);
+    }
+
+    @Override
+    public boolean exists(URI file)
+    {
+        return getS3ObjectMetadata(file) != null;
+    }
+
+    @Override
+    public void createFile(URI file)
+            throws IOException
+    {
+        s3.putObject(getBucketName(file), keyFromUri(file), "");
+    }
+
+    @Override
+    public void deleteRecursively(URI uri)
+            throws IOException
+    {
+        if (!isDirectory(uri)) {
+            s3.deleteObject(getBucketName(uri), keyFromUri(uri));
+        }
+        else {
+            listDirectories(uri).forEach(dir -> {
+                try {
+                    deleteRecursively(dir);
+                }
+                catch (IOException e) {
+                    e.printStackTrace();
+                }
+            });
+            listFiles(uri).forEach(file -> s3.deleteObject(getBucketName(file), keyFromUri(file)));
+            s3.deleteObject(getBucketName(uri), keyFromUri(uri) + DIRECTORY_SUFFIX);
+        }
+    }
+
+    @Override
+    public Stream<URI> listDirectories(URI dir)
+    {
+        List<String> prefixes = listS3Directory(dir).getCommonPrefixes();
+        return prefixes.stream().map(prefix -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + prefix, dir.getFragment());
+            }
+            catch (URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        });
+    }
+
+    @Override
+    public Stream<URI> listFiles(URI dir)
+    {
+        List<S3ObjectSummary> objects = listS3Directory(dir).getObjectSummaries();
+        return objects.stream().filter(object -> !object.getKey().endsWith(PATH_SEPARATOR)).map(object -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + object.getKey(), dir.getFragment());
+            }
+            catch (URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        });
+    }
+
+    @Override
+    public long size(URI file)
+            throws IOException
+    {
+        return getS3ObjectMetadata(file).getContentLength();
+    }
+
+    @Override
+    public void close()
+    {
+        uploadExecutor.shutdown();
+        s3.shutdown();
```

Comment:
Maybe not important here, but I prefer to use `Closer` in `close()` implementations when we need to close multiple building blocks. So we are sure that exception when closing one of building blocks does not prevent us from closing ohters.